cli/pull.go

```diff
@@ -228,3 +210,54 @@ IMAGE ID            IMAGE NAME                           SIZE
 bbc3a0323522        docker.io/library/busybox:latest     703.14 KB
 0153c5db97e5        docker.io/library/redis:alpine       9.63 MB`
 }
+
+// pullMissingImage pull the image if it doesn't exist.
+// When `force` is true, always pull the latest image instead of
+// using the local version
+func pullMissingImage(ctx context.Context, apiClient client.CommonAPIClient, image string, force bool) error {
+	var inspectError error
+	if !force {
+		_, inspectError = apiClient.ImageInspect(ctx, image)
+		if inspectError == nil {
+			return nil
+		}
+	}
+
+	namedRef, err := reference.Parse(image)
+	if err != nil {
+		return fmt.Errorf("failed to parse image reference: %v", err)
+	}
+	namedRef = reference.TrimTagForDigest(reference.WithDefaultTagIfMissing(namedRef))
+
+	// the error comes from daemon/mgr/image_store.go
```

Comment:
Hi, @wrfly , I think we can use assert to check the error is `RespError` or not.  If the statusCode is 404, we can continue to pull the image. It seems that it's better than using string match. WDYT?

Reply:
Yeah, assert is better, I just use it in testing before. I'll change it later.

Reply:
但是返回的 `inspectError` 就是一个单纯的error呀，没有 `statusCode`， CommonAPIClient 把 `resp.StatusCode` 吃掉了

```go

// ImageInspect requests daemon to inspect an image.
func (client *APIClient) ImageInspect(ctx context.Context, name string) (types.ImageInfo, error) {
	image := types.ImageInfo{}

	resp, err := client.get(ctx, "/images/"+name+"/json", nil, nil)
	if err != nil {
		return image, err
	}

	defer ensureCloseReader(resp)
	err = decodeBody(&image, resp.Body)
	return image, err
}
```

Reply:
Hi @wrfly , please check the `client.get` method and it will return `RespError`. 

you can use the following statement to check the error:

```
if eErr, ok := err.(client.RespError); ok {
     // we need to add StatusCode method. 
     if e.Err.StatusCode() != 404 {
     }
}
```

Does it make senses to you?

@wrfly please try to use English. :) Try your best

Reply:
Yeah, understood.

I thought it's ugly to convert errors around at first, but I saw `moby` use it too. It's my problem that I haven't see this usage before.

Reply:
OK.  So please update your code 😄

Reply:
updated